core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -689,6 +689,64 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
     new ParallelCollectionRDD[T](this, seq, numSlices, Map[Int, Seq[String]]())
   }
 
+  /**
+   * Creates a new RDD[Long] containing elements from `start` to `end`(exclusive), increased by
+   * `step` every element.
+   *
+   * @note if we need to cache this RDD, we should make sure each partition contains no more than
+   *       2 billion element.
+   *
+   * @param start the start value.
+   * @param end the end value.
+   * @param step the incremental step
+   * @param numSlices the partition number of the new RDD.
+   * @return
+   */
+  def range(
+      start: Long,
+      end: Long,
+      step: Long = 1,
+      numSlices: Int = defaultParallelism): RDD[Long] = withScope {
+    assertNotStopped()
+    if (step == 0) {
+      // when step is 0, range will run infinite
+      throw new IllegalArgumentException("`step` cannot be 0")
+    }
+    val length =
+      if ((end - start) % step == 0) {
+        (end - start) / step
+      } else {
+        (end - start) / step + 1
+      }
+    parallelize(0 to numSlices).mapPartitions(iter => {
+      val i = iter.next()
+      val partitionStart = (i * length) / numSlices * step + start
+      val partitionEnd = ((i + 1) * length) / numSlices * step + start
```

Comment:
After you update to `parallelize(0 until numSlices, numSlices)`, please try this case: `range(0, 10, 1, 3)`. length = 10, numSlices = 1. So the there partitions will be [0, 3), [3, 6), [6, 9). `9` is lost. It works because you use `parallelize(0 until numSlices)`, which creates `numSlices + 1` partitions.